File: code/code cpp/codi_alba/interpolacio.cpp
#include "interpolacio.h"
#include <cmath>

void interpolacio(const std::vector<double>& x, const std::vector<double>& Q,
                  double SNR, double R, double dE_0, double dE_1,
                  double& rho, double& a, double& b, double& c) {
    double sum = 0.0;
    int len = x.size();
    for (int i = 0; i < len; ++i) {
        for (int j = 0; j < len; ++j) {
            double diff = x[i] - x[j];
            sum += Q[i] * Q[j] * std::exp(-SNR * (diff * diff) / 4.0);
        }
    }

    double f2 = -std::log2(sum) - R;
    double dp1 = dE_0 - R;
    double dp2 = dE_1 - R;

    a = 3.0 * (dp2 + dp1 - 2.0 * f2);
    b = 2.0 * (3.0 * f2 - 2.0 * dp1 - dp2);
    c = dp1;

    double discriminant = b * b - 4.0 * a * c;
    if (discriminant < 0) {
        rho = 0; // o algún valor por defecto
    } else {
        rho = (-b - std::sqrt(discriminant)) / (2.0 * a);
    }
}

File: code/exponents_matlab/interpolacio.m
function [rho, a, b, c] = interpolacio(x, Q, SNR, R, dE_0, dE_1)
% IINTERPOLACIÓ - Hermites
% E0(1)
sum = 0;
for i = 1:length(x)
    for j = 1:length(x)
       sum = sum + Q(i)*Q(j)*exp(-SNR*((x(i)-x(j))^2)/4);
    end
end

f2 = -log2(sum)-R;
dp1 = dE_0 - R;
dp2 = dE_1 - R;

% Coeficients a, b, c de la derivada: a x^2 + b x + c
a = 3*(dp2 + dp1 - 2*f2);
b = 2*(3*f2 - 2*dp1 - dp2);
c = dp1;

% Arrels de la derivada
%rho = (-b + sqrt(b^2 - 4*a*c)) / (2*a)
rho = (-b - sqrt(b^2 - 4*a*c)) / (2*a);  % No estem 100% de que sigui aquesta pero quien no arriesga no gana
end